### ISSUE TITLE
-O1 Optimization failure issue

### ISSUE THREAD

**Issue by petrokarashchenko** [foss-for-synopsys-dwc-arc-processors/toolchain#232]
I have an issue. I have a project that I try to compile for arc600 BE CPU.
There are quite many files in the project and I observe the next issue:
When I compile code with `-O1` I observe strange behaviour in my code. I have not found the root cause, so it is still in progress.
When I compile code with `-O0`,  then all is working as expected.

With this I have evaluated then either that my code has some fault or there is an issue with optimization (similar to https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/issues/220).

So I tried to add all optimization added by `-O1` manually and see when error start reproducing.
To do this I have used next options to find out optimization features enabled by `-O1`. I used `-fverbose-asm` option. Then I copied all applied options and replaced `-O1` option. As a result my code is working as expected.

So I have a question: What options are actually applied with `-O1` when compile for ARC? Are there are "hidden" optimisations? And how to enable/disable those?

I have used
```
-faggressive-loop-optimizations \
-fauto-inc-dec \
-fbranch-count-reg \
-fcombine-stack-adjustments \
-fcompare-elim \
-fcprop-registers \
-fdefer-pop \
-fdelayed-branch \
-fdelete-null-pointer-checks \
-fdwarf2-cfi-asm \
-fearly-inlining \
-feliminate-unused-debug-types \
-fforward-propagate \
-ffp-int-builtin-inexact \
-ffunction-cse \
-fgcse-lm \
-fgnu-unique \
-fguess-branch-probability \
-fident \
-fif-conversion \
-fif-conversion2 \
-finline-atomics \
-finline-functions-called-once \
-fipa-profile \
-fipa-pure-const \
-fipa-reference \
-fipa-reference-addressable \
-fipa-stack-alignment \
-fira-hoist-pressure \
-fira-share-save-slots \
-fira-share-spill-slots \
-fivopts \
-fkeep-static-consts \
-flifetime-dse \
-flto-odr-type-merging \
-fmath-errno \
-fmerge-constants \
-fmerge-debug-strings \
-fmove-loop-invariants \
-fomit-frame-pointer \
-fpeephole \
-fplt \
-freg-struct-return \
-freorder-blocks \
-fsched-critical-path-heuristic \
-fsched-dep-count-heuristic \
-fsched-group-heuristic \
-fsched-interblock \
-fsched-last-insn-heuristic \
-fsched-rank-heuristic \
-fsched-spec \
-fsched-spec-insn-heuristic -fsched-stalled-insns-dep -fschedule-fusion \
-fsemantic-interposition -fshow-column -fshrink-wrap \
-fshrink-wrap-separate -fsigned-zeros -fsplit-ivs-in-unroller \
-fsplit-wide-types -fssa-backprop -fssa-phiopt -fstdarg-opt \
-fstrict-volatile-bitfields -fsync-libcalls -ftoplevel-reorder \
-ftrapping-math -ftree-bit-ccp -ftree-builtin-call-dce -ftree-ccp \
-ftree-ch -ftree-coalesce-vars -ftree-copy-prop -ftree-cselim -ftree-dce \
-ftree-dominator-opts -ftree-dse -ftree-forwprop -ftree-fre \
-ftree-loop-if-convert -ftree-loop-im -ftree-loop-ivcanon \
-ftree-loop-optimize -ftree-parallelize-loops=0 -ftree-phiprop -ftree-pta \
-ftree-reassoc -ftree-scev-cprop -ftree-sink -ftree-slsr -ftree-sra \
-ftree-ter -funit-at-a-time -fzero-initialized-in-bss
```
vs
```
-O1 -fno-crossjumping
```

Case with `-O1` generates slightly different code compared to option set.

Could you please help me to find a root cause?

**Comment by petrokarashchenko** [comment]
I have tried to use option set for `-O1` specified in https://gcc.gnu.org/onlinedocs/gcc-9.2.0/gcc/Optimize-Options.html
with
```
-fauto-inc-dec 
-fbranch-count-reg 
-fcombine-stack-adjustments 
-fcompare-elim 
-fcprop-registers 
-fdce 
-fdefer-pop 
-fdelayed-branch 
-fdse 
-fforward-propagate 
-fguess-branch-probability 
-fif-conversion 
-fif-conversion2 
-finline-functions-called-once 
-fipa-profile 
-fipa-pure-const 
-fipa-reference 
-fipa-reference-addressable 
-fmerge-constants 
-fmove-loop-invariants 
-fomit-frame-pointer 
-freorder-blocks 
-fshrink-wrap 
-fshrink-wrap-separate 
-fsplit-wide-types 
-fssa-backprop 
-fssa-phiopt 
-ftree-bit-ccp 
-ftree-ccp 
-ftree-ch 
-ftree-coalesce-vars 
-ftree-copy-prop 
-ftree-dce 
-ftree-dominator-opts 
-ftree-dse 
-ftree-forwprop 
-ftree-fre 
-ftree-phiprop 
-ftree-pta 
-ftree-scev-cprop 
-ftree-sink 
-ftree-slsr 
-ftree-sra 
-ftree-ter 
-funit-at-a-time
```
My code also works as expected

**Comment by claziss** [comment]
You will get the #220 issue whenever you use either some fancy CFG optimization or if-convertion (insn predication). Probably now you have something related to if-conversion steps. You can try to turn them off.

**Comment by petrokarashchenko** [comment]
@claziss I think I have found the root cause of the issue. At least some assembly generated code that seems to be buggy to me.

I have next code snippet:
```
typedef volatile struct {
    uint8_t  *a;
    uint8_t  *b;
    uint8_t  *c;
} __attribute__((uncached)) my_data_t;

typedef volatile struct {
    my_data_t  data;
} __attribute__((uncached)) my_type_t;

void foo (my_type_t *p, uint32_t val)
{
    if ((p->data.c + val) >= p->data.b) {
        p->data.c = p->data.a + (p->data.c + val - p->data.b);
    } else {
        p->data.c += val;
    }
}
```

When I compile it with `-O0` I get next assembly that is working as expected:
```
  12               		.global	foo
  14               	foo:
  15 0000 1CFC B6C8 		st.a	fp,[sp,-4]	;26
  16 0004 230A 3700 		mov	fp,sp	;5
  17 0008 C1A2      		sub_s sp,sp,8 ;4
  18 000a 1BFC B000 		st	r0,[fp,-4]	;26
  19 000e 1BF8 B040 		st	r1,[fp,-8]	;26
  20 0012 13FC B002 		ld	r2,[fp,-4]	;23
  21 0016 1208 0803 		ld.di	r3,[r2,8]	;23
  22 001a 13F8 B002 		ld	r2,[fp,-8]	;23
  23 001e 635B      		add_s r3,r3,r2 ;1
  24 0020 13FC B002 		ld	r2,[fp,-4]	;23
  25 0024 1204 0802 		ld.di	r2,[r2,4]	;23
  26 0028 7350      		cmp_s r3,r2
  27 002a F79A      		blo_s @.L2
  28 002c 13FC B002 		ld	r2,[fp,-4]	;23
  29 0030 1200 0802 		ld.di	r2,[r2]	;23
  30 0034 13FC B003 		ld	r3,[fp,-4]	;23
  31 0038 1308 080C 		ld.di	r12,[r3,8]	;23
  32 003c 13F8 B003 		ld	r3,[fp,-8]	;23
  33 0040 647C      		add_s r12,r12,r3 ;1
  34 0042 13FC B003 		ld	r3,[fp,-4]	;23
  35 0046 1304 0803 		ld.di	r3,[r3,4]	;23
  36 004a 230E 0303 		rsub r3,r3,r12
  37 004e 627B      		add_s r3,r2,r3 ;1
  38 0050 13FC B002 		ld	r2,[fp,-4]	;23
  39 0054 1A08 00E0 		st.di	r3,[r2,8]	;26
  40 0058 F00D      		b_s @.L4
  41 005a 78E0      		.align 4
  42               	.L2:
  43 005c 13FC B002 		ld	r2,[fp,-4]	;23
  44 0060 1208 0803 		ld.di	r3,[r2,8]	;23
  45 0064 13F8 B002 		ld	r2,[fp,-8]	;23
  46 0068 635B      		add_s r3,r3,r2 ;1
  47 006a 13FC B002 		ld	r2,[fp,-4]	;23
  48 006e 1A08 00E0 		st.di	r3,[r2,8]	;26
  49               		.align 2
  50               	.L4:
  51 0072 78E0      		nop_s
  52 0074 240A 36C0 		mov	sp,fp	;5
  53 0078 1404 341B 		ld.ab	fp,[sp,4]	;23
  54 007c 7EE0      		j_s	[blink]
```

 But when I compile same code with `-Os -fno-crossjumping` I'm getting next assembly that does not work as expected:
```
  12               		.global	foo
  14               	foo:
  15 0000 1008 0802 		ld.di	r2,[r0,8]	;23
  16 0004 1004 0803 		ld.di	r3,[r0,4]	;23
  17 0008 623A      		add_s r2,r2,r1 ;1
  18 000a 0A1D 00C4 		brhi r3, r2, @.L2
  19 000e 1000 0803 		ld.di	r3,[r0]	;23
  20 0012 1008 0802 		ld.di	r2,[r0,8]	;23
  21 0016 1004 080C 		ld.di	r12,[r0,4]	;23
  22 001a 6239      		add_s r1,r2,r1 ;1
  23 001c 7982      		sub_s r1,r1,r12
  24 001e 6339      		add_s r1,r3,r1 ;1
  25 0020 A022      		st_s	r1,[r0,8]		;16
  26 0022 7EE0      		j_s	[blink]
  27               		.align 2
  28               	.L2:
  29 0024 1008 0802 		ld.di	r2,[r0,8]	;23
  30 0028 6239      		add_s r1,r2,r1 ;1
  31 002a A022      		st_s	r1,[r0,8]		;16
  32 002c 7EE0      		j_s	[blink]
```

I'm getting `st_s	r1,[r0,8]` instead of `st.di	r1,[r0,8]` and `st_s	r1,[r0,8]` instead of `st.di	r1,[r0,8]` in both cases just before `j_s	[blink]` instruction.

Could you please take a look into this issues as well?

**Comment by petrokarashchenko** [comment]
In this particular code snipped the workaround is to use `-fno-if-conversion2`. It is very strange, but code with `-fif-conversion2` is:
```
  14               	foo:
  15 0000 1008 0802 		ld.di	r2,[r0,8]	;23
  16 0004 1004 0803 		ld.di	r3,[r0,4]	;23
  17 0008 623A      		add_s r2,r2,r1 ;1
  18 000a 0A1D 00C4 		brhi r3, r2, @.L2
  19 000e 1000 0803 		ld.di	r3,[r0]	;23
  20 0012 1008 0802 		ld.di	r2,[r0,8]	;23
  21 0016 1004 080C 		ld.di	r12,[r0,4]	;23
  22 001a 6239      		add_s r1,r2,r1 ;1
  23 001c 7982      		sub_s r1,r1,r12
  24 001e 6339      		add_s r1,r3,r1 ;1
  25 0020 A022      		st_s	r1,[r0,8]		;16
  26 0022 7EE0      		j_s	[blink]
  27               		.align 2
  28               	.L2:
  29 0024 1008 0802 		ld.di	r2,[r0,8]	;23
  30 0028 6239      		add_s r1,r2,r1 ;1
  31 002a A022      		st_s	r1,[r0,8]		;16
  32 002c 7EE0      		j_s	[blink]
```
and with `-fno-if-conversion2` is
```
  14               	foo:
  15 0000 1008 0802 		ld.di	r2,[r0,8]	;23
  16 0004 1004 0803 		ld.di	r3,[r0,4]	;23
  17 0008 623A      		add_s r2,r2,r1 ;1
  18 000a 0A1F 00C4 		brhi r3, r2, @.L2
  19 000e 1000 0803 		ld.di	r3,[r0]	;23
  20 0012 1008 0802 		ld.di	r2,[r0,8]	;23
  21 0016 1004 080C 		ld.di	r12,[r0,4]	;23
  22 001a 6239      		add_s r1,r2,r1 ;1
  23 001c 7982      		sub_s r1,r1,r12
  24 001e 6339      		add_s r1,r3,r1 ;1
  25 0020 1808 0060 		st.di	r1,[r0,8]	;26
  26 0024 7EE0      		j_s	[blink]
  27               		.align 2
  28               	.L2:
  29 0026 1008 0802 		ld.di	r2,[r0,8]	;23
  30 002a 6239      		add_s r1,r2,r1 ;1
  31 002c 1808 0060 		st.di	r1,[r0,8]	;26
  32 0030 7EE0      		j_s	[blink]
```
so difference is only in `st_s	r1,[r0,8]` vs `st.di	r1,[r0,8]`.

I have disabled `if-conversion2`, but now my code does not work in another places. Will look for further code snippets that have `.di` discarded due to optimization

**Comment by petrokarashchenko** [comment]
Using next options makes code works as expected.
```
-fno-tree-bit-ccp -fno-tree-builtin-call-dce -fno-tree-ccp \
-fno-tree-ch -fno-tree-coalesce-vars -fno-tree-copy-prop -fno-tree-cselim -fno-tree-dce \
-fno-tree-dominator-opts -fno-tree-dse -fno-tree-forwprop -fno-tree-fre \
-fno-tree-loop-if-convert -fno-tree-loop-im -fno-tree-loop-ivcanon \
-fno-tree-loop-optimize -fno-tree-phiprop -fno-tree-pta \
-fno-tree-reassoc -fno-tree-scev-cprop -fno-tree-sink -fno-tree-slsr -fno-tree-sra \
-fno-tree-ter
```
Just a backbox test. But need to compare assembly further for code samples.

**Comment by claziss** [comment]
I have something in crafting process. I may be able to release it somewhere next week in the first part.

**Comment by petrokarashchenko** [comment]
@claziss Thanks. Looking forward to try you craft results ;)

**Comment by claziss** [comment]
I've made a new approach for direct access. It may not be as optimal as previous, but it should conserve the _uncached_ attribute until code gen. Patch here:
https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/commit/2080016aa654d7d4b8dc4fddc1e151d5ffa35d54

**Comment by petrokarashchenko** [comment]
All my issues related to `uncached` attribute are fixed using latest commits from `2020.03` gcc branch

**Comment by claziss** [comment]
@petrokarashchenko I've rebased/merged/fixed the 2030.03. Please refer to the latest. I'll close this issue, and keep only one.

**Comment by abrodkin** [comment]
Fixed by https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/commit/fc5aa9c7be10d21206537161f2245c097e06f456 in `arc-2020.03`.
